### PR TITLE
Deployment: filter out old RSes that are deleted or with non-zero replicas before cleanup

### LINKED ISSUE #36379: [k8s.io] Deployment deployment should delete old replica sets {Kubernetes e2e suite}
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-e2e-gci-gke-release-1.4/2516/

Failed: [k8s.io] Deployment deployment should delete old replica sets {Kubernetes e2e suite}

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:68
Expected error:
    <errors.aggregate | len:1, cap:1>: [
        {
            ErrStatus: {
                TypeMeta: {Kind: "", APIVersion: ""},
                ListMeta: {SelfLink: "", ResourceVersion: ""},
                Status: "Failure",
                Message: "an error on the server (\"Internal Server Error: \\\"/apis/extensions/v1beta1/namespaces/e2e-tests-deployment-tklh9/replicasets/test-cleanup-deployment-2963253163\\\"\") has prevented the request from succeeding (get replicasets.extensions test-cleanup-deployment-2963253163)",
                Reason: "InternalError",
                Details: {
                    Name: "test-cleanup-deployment-2963253163",
                    Group: "extensions",
                    Kind: "replicasets",
                    Causes: [
                        {
                            Type: "UnexpectedServerResponse",
                            Message: "Internal Server Error: \"/apis/extensions/v1beta1/namespaces/e2e-tests-deployment-tklh9/replicasets/test-cleanup-deployment-2963253163\"",
                            Field: "",
                        },
                    ],
                    RetryAfterSeconds: 0,
                },
                Code: 500,
            },
        },
    ]
    an error on the server ("Internal Server Error: \"/apis/extensions/v1beta1/namespaces/e2e-tests-deployment-tklh9/replicasets/test-cleanup-deployment-2963253163\"") has prevented the request from succeeding (get replicasets.extensions test-cleanup-deployment-2963253163)
not to have occurred
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:202
```

Previous issues for this test: #28339

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #36379

cc @zmerlynn @yujuhong @kargakis @kubernetes/sig-apps-bugs

### CODE REVIEW

**Comment by yujuhong** [comment]
Looks reasonable to me. I'll leave the detailed code review to the package owners.

**Comment by yujuhong** [comment]
The excessive "delete" requests may have caused conflicts with the garbage collector, delaying the actual deletion of the replica set. The PR fixes that and could help unblock the submit queue (ci-kubernetes-e2e-kops-aws failing). Applying the priority label accordingly. /cc @k8s-oncall 

Another thing is that the garbage collector should log the error at a lower level. Now it's using V(6) and that's not helpful at all....

**Comment by yujuhong** [comment]
LGTM

**Comment by kargakis** [comment]
@janetkuo I already have a PR in-flight that does exactly what you are doing here albeit it has only the minimal diff required: https://github.com/kubernetes/kubernetes/pull/41145

**Comment by janetkuo** [comment]
> I already have a PR in-flight that does exactly what you are doing here albeit it has only the minimal diff required

The fix is slightly different since I intentionally move the old RS filter to before calculating `diff`. I made a comment in #41145

File: pkg/controller/controller_utils.go [hunk 1]
```diff
@@ -754,15 +754,21 @@ func IsPodActive(p *v1.Pod) bool {
...
+	activeFilter := func(rs *extensions.ReplicaSet) bool {
+		return rs != nil && *(rs.Spec.Replicas) > 0
+	}
+	return FilteredRSes(replicaSets, activeFilter)
+}
 
-		if rs != nil && *(rs.Spec.Replicas) > 0 {
-			active = append(active, replicaSets[i])
+// FilteredRSes returns replica sets that are filtered by filterFn (all returned ones should match filterFn).
+func FilteredRSes(RSes []*extensions.ReplicaSet, filterFn func(rs *extensions.ReplicaSet) bool) []*extensions.ReplicaSet {
```
Comment (kargakis) [inline pkg/controller/controller_utils.go]: FilterReplicaSets

Comment (kargakis) [inline pkg/controller/controller_utils.go]: Also make filterFn a type

**Comment by kargakis** [comment]
Can you add unit tests for the filter functions? I don't think they have any...

**Comment by janetkuo** [comment]
@kargakis ptal

File: pkg/controller/controller_utils.go (hunk 1 above)
Comment (janetkuo) [inline pkg/controller/controller_utils.go]: Done

File: pkg/controller/deployment/sync.go [hunk 2]
```diff
@@ -541,22 +541,25 @@ func (dc *DeploymentController) cleanupDeployment(oldRSs []*extensions.ReplicaSe
...
 	}
-	diff := int32(len(oldRSs)) - *deployment.Spec.RevisionHistoryLimit
+
+	// Avoid deleting replica set with deletion timestamp set or with non-zero replica counts
+	cleanablefilter := func(rs *extensions.ReplicaSet) bool {
+		return rs != nil && rs.ObjectMeta.DeletionTimestamp == nil && rs.Status.Replicas == 0 && *(rs.Spec.Replicas) == 0 && rs.Generation <= rs.Status.ObservedGeneration
+	}
+	cleanableRSes := controller.FilterReplicaSets(oldRSs, cleanablefilter)
+
+	diff := int32(len(cleanableRSes)) - *deployment.Spec.RevisionHistoryLimit
```
Comment (kargakis) [inline pkg/controller/deployment/sync.go]: I think this slightly changes the meaning of revisionHistoryLimit. If I have a Deployment (revisionHistoryLimit=1) with two old ReplicaSets and a new one:
d.replicas=3 with rs-1.replicas=0 (old), rs-2.replicas=1 (old), and rs-3.replicas=2 (new)

With the old behavior we would cleanup rs-1, with the behavior you have here we won't until the rollout finishes. I think revisionHistoryLimit should take all old controllers into account and then filter out those it cannot delete as we already do. Maybe prefiltering just for deletionTimestamp is enough.

Comment (janetkuo) [inline pkg/controller/deployment/sync.go]: agree

Comment (janetkuo) [inline pkg/controller/deployment/sync.go]: Done

File: pkg/controller/deployment/sync.go [hunk 3]
```diff
@@ -541,20 +541,27 @@ func (dc *DeploymentController) cleanupDeployment(oldRSs []*extensions.ReplicaSe
...
+	sort.Sort(controller.ReplicaSetsByCreationTimestamp(cleanableRSes))
 	glog.V(2).Infof("Looking to cleanup old replica sets for deployment %q", deployment.Name)
 
 	var errList []error
 	// TODO: This should be parallelized.
 	for i := int32(0); i < diff; i++ {
-		rs := oldRSs[i]
+		rs := cleanableRSes[i]
 		// Avoid delete replica set with non-zero replica counts
-		if rs.Status.Replicas != 0 || *(rs.Spec.Replicas) != 0 || rs.Generation > rs.Status.ObservedGeneration || rs.DeletionTimestamp != nil {
```
Comment (kargakis) [inline pkg/controller/deployment/sync.go]: No reason to update this line, also leave the deletionTimestamp check as a last line of defense.

**Comment by kargakis** [comment]
One last comment, lgtm otherwise

File: pkg/controller/deployment/sync.go (hunk 3 above)
Comment (janetkuo) [inline pkg/controller/deployment/sync.go]: Ah, thanks, it's a rebase mistake

**Comment by janetkuo** [comment]
@k8s-bot kubemark e2e test this

**Comment by janetkuo** [comment]
Applying tag since got LGTM'ed

**Pull request merged**